examples/generic_events.rs

```diff
@@ -39,12 +39,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     // Ask for present ConfigureNotify events
     let event_id = conn.generate_id()?;
-    present::select_input(
-        &conn,
-        event_id,
-        win_id,
-        present::EventMask::ConfigureNotify.into(),
-    )?;
+    present::select_input(&conn, event_id, win_id, present::EventMask::ConfigureNotify)?;
```

Comment:
As I commented in the `send_event` case, an `Into<>` bound is used for every `enum`(when the enum type cannot be used directly)/`altenum`/`mask`/`altmask`.

Reply:
Why the "when the enum type cannot be used directly" restriction? In that other comment, you mentioned duplicate values. Why are they a problem? I'd expect this to just always be possible when a mask/altmask/enum/altenum property is present.

Reply:
Because in that case, the argument type is directly the enum (like in structs after https://github.com/psychon/x11rb/pull/299)

Reply:
Sorry, I may not see the obvious, but... shouldn't it always be possible to have the enum directly as argument for fields with `enum`? It does not matter if an enum has duplicate values for this or not.

So, to summarise, may understanding currently is:

- for `enum` arguments, one can just always have the enum type as argument
- for `altenum`, `mask`, `altmask` arguments, it makes sense to use `Into<uX>` for the argument (so that it becomes possible to directly pass in enum variants)
- for arguments without `enum`/`altenum`/`mask`/`altmask` properties, the argument type should be the "plain" type (e.g. `u32`)

The second item is new and a good idea and not yet part of the current code generator.

Reply:
Sometimes, one may want to copy the value of the enum from a struct to pass it as argument. If the enum has repeated values, the struct will have the "plain" and passing it directly as argument wouldn't work if the argument only accepts the enum type.

Reply:
The only such case that I know about is `xproto::Gravity`:
https://github.com/psychon/x11rb/blob/34014bb7763a797374464301b868584a9559efe3/code_generator_helpers/module.py#L936-L948
This only appears in `GetWindowAttributesReply`:
https://github.com/psychon/x11rb/blob/34014bb7763a797374464301b868584a9559efe3/xcb-proto-1.14-1-g2b3559c/src/xproto.xml#L1562-L1563
The FIXME in the first code sample is about handling this: In `bit_gravity`, value 0 needs to map to `BitForget` while for `win_gravity` that is `WinUnmap`.

No idea yet how to implement this, but once that's done, I do know about any cases where an "unparsed" enum appears in some kind of structure (neither reply, not event, not error).

So, assuming I did not forget anything, does that mean the special handling for enums with duplicate values is not necessary here?

Reply:
Oh, fun fact: I cannot actually do this in the current code generator since `xcbgen` does not fetch anything but `enum` from the XML.

Reply:
I will think about this later.



> Oh, fun fact: I cannot actually do this in the current code generator since `xcbgen` does not fetch anything but `enum` from the XML.

I have the feeling that the Python xcbgen is very `c_client.py` oriented.

Reply:
> I have the feeling that the Python xcbgen is very c_client.py oriented.

What could you mean? Are there any places where it assumes that one generates C code? Perhaps even something that made it difficult to handle `BOOL` specially? ;-)

https://github.com/psychon/x11rb/blob/b846f205b522f3f2e801f795729ba81ec3ea7027/xcb-proto-1.14-1-g2b3559c/xcbgen/state.py#L94-L108